Étape 4 — Recommandation initiale de parcours

recommande le parcours à partir du test initial.

# Cellule 1 — Imports

In [1]:
# ============================================================
# ÉTAPE 4 — RECOMMANDATION INITIALE DE PARCOURS
# ============================================================

import pandas as pd
import numpy as np
from pathlib import Path
import json
from datetime import datetime

pd.set_option("display.max_columns", 200)
pd.set_option("display.max_rows", 100)

# Cellule 2 — Chemins et chargement

In [2]:
# ============================================================
# 4.1. Chargement du dataset d'orientation scoré
# ============================================================

INPUT_DIR = Path("outputs/02_scoring_pedagogique")
OUTPUT_DIR = Path("outputs/04_recommandation_parcours")
OUTPUT_DIR.mkdir(parents=True, exist_ok=True)

PATH_ORIENTATION_SCORE = INPUT_DIR / "04_dataset_orientation_score.csv"

df_orientation = pd.read_csv(PATH_ORIENTATION_SCORE)

print("Dataset orientation chargé :", df_orientation.shape)
display(df_orientation.head())

Dataset orientation chargé : (305, 59)


,ins_CreatedAt,IDENTIFICATION,ins_Année de naissance,ins_Genre,ins_Région d'origine,ins_Filière,ins_Niveau d'étude,ins_Statut actuel,ins_Niveau informatique,ins_Niveau Excel,ins_Niveau Power BI,ins_Niveau Python,ins_Niveau IA,ins_Motivation,ins_Objectif pro,ins_Disponibilité,ins_Source info,ins_annee_naissance_originale,ins_annee_naissance,ins_age,init_CreatedAt,init_Date du test,init_Q1 - Resume ventes Excel,init_Q2 - Import CSV Excel,init_Q3 - Modele donnees Excel,init_Q4 - Mauvaise pratique visu Excel,init_Q5 - 2e grande valeur Excel,init_Q6 - Mesure vs Colonne DAX,init_Q7 - CA annee precedente DAX,init_Q8 - Vue Modele Power BI,init_Q9 - Acces directeurs regionaux,init_Q10 - 12 commerciaux 3 indicateurs,init_Q11 - Bibliotheque CSV Python,init_Q12 - Overfitting Underfitting,init_Q13 - Segmentation 50000 clients,init_Q14 - Deployer modele Python API,init_Q15 - Valeurs manquantes 30pc,init_Q16 - Role system prompt LLM,init_Q17 - Assistant IA PDF financiers,init_Q18 - Role embedding dans RAG,init_Q19 - Agent IA selection outil,init_Q20 - Sortie fiable LLM tableau,score_initial_DA,score_initial_DA_pct,niveau_initial_DA,score_initial_BI,score_initial_BI_pct,niveau_initial_BI,score_initial_DS,score_initial_DS_pct,niveau_initial_DS,score_initial_IA,score_initial_IA_pct,niveau_initial_IA,score_initial_total,score_initial_pourcentage,niveau_initial_global,unknown_total_initial,cluster_initial
0,2026-03-31 23:34:20+00:00,DIEG2026-001,2007,Masculin,Analamanga,Informatique / Genie logiciel,Licence 2 (L2),Etudiant(e) a plein temps,Intermediaire,Avance,Intermediaire,Intermediaire,Debutant,"[""Evoluer dans mon poste""]",Data Scientist,"[""En semaine - Matin""]",LinkedIn,2007,2007.0,19.0,2026-03-31 23:36:12+00:00,2026-03-31 23:36:11+00:00,A. Filtre automatique,C. Formules SI imbriquees,D. Formules RECHERCHEV,B. Limiter les couleurs,D. RANG(plage;2),C. Mesure=visualisation Colonne=stockee,B. CALCULATE(SUM SAMEPERIODLASTYEAR),D. Definir relations entre tables,C. Rapport separe par directeur,D. Radar ou tableau matriciel conditionnel,E. Je ne sais pas,C. Underfitting,C. Regression lineaire,C. FastAPI ou Flask + joblib ou pickle,D. Supprimer toutes les lignes,E. Je ne sais pas,B. RAG - indexer et recuperer dynamiquement,C. Compresser les fichiers PDF,B. Ordre dans lequel les outils sont listes,C. Augmenter la temperature,0,0.0,Débutant,4,80.0,Avancé,1,20.0,Débutant,1,20.0,Débutant,6,30.0,Débutant,2,Débutant
1,2026-04-01 00:36:16+00:00,DIEG2026-002,20002,Masculin,Atsimo-Atsinanana,Polytechnique / Genie civil / Genie industriel,Master 1 (M1),Diplome(e) en recherche d emploi,Avance,Avance,Avance,Avance,Intermediaire,"[""Acquerir des competences techniques""]",Developpeur Python / IA,"[""En semaine - Apres-midi""]",Ami(e),20002,NaN,NaN,2026-04-01 00:37:59+00:00,2026-04-01 00:38:00+00:00,C. Formule SOMME(),A. Macros VBA manuelles,C. Power Pivot avec relations,C. Etiquettes de donnees claires,C. MAX(SI(...)),C. Mesure=visualisation Colonne=stockee,E. Je ne sais pas,B. Publier sur Power BI Service,B. Row-Level Security (RLS),B. Graphique en secteurs,A. matplotlib,C. Underfitting,C. Regression lineaire,D. Envoyer par e-mail,D. Supprimer toutes les lignes,E. Je ne sais pas,B. RAG - indexer et recuperer dynamiquement,B. Authentifier utilisateur,E. Je ne sais pas,D. JSON structure avec schema defini,1,20.0,Débutant,2,40.0,Débutant,0,0.0,Débutant,2,40.0,Débutant,5,25.0,Débutant,3,Débutant
2,2026-04-01 03:01:14+00:00,DIEG2026-003,2001,Feminin,Diana,Droit / Sciences politiques,Master 2 (M2),Diplome(e) en recherche d emploi,Aucun,Aucun,Aucun,Aucun,Aucun,"[""Ameliorer mon employabilite"", "" Acquerir des...",Entrepreneur,"[""En semaine - Matin""]",Facebook,2001,2001.0,25.0,2026-04-03 13:28:48+00:00,2026-04-03 13:28:47+00:00,E. Je ne sais pas,E. Je ne sais pas,E. Je ne sais pas,E. Je ne sais pas,E. Je ne sais pas,E. Je ne sais pas,E. Je ne sais pas,E. Je ne sais pas,E. Je ne sais pas,E. Je ne sais pas,E. Je ne sais pas,E. Je ne sais pas,E. Je ne sais pas,E. Je 

# Cellule 3 — Paramètres de recommandation

In [3]:
# ============================================================
# 4.2. Paramètres pédagogiques
# ============================================================

SEUIL_DEBUTANT = 50
SEUIL_INTERMEDIAIRE = 70

domaines = ["DA", "BI", "DS", "IA"]

ordre_parcours = ["DA", "BI", "DS", "IA"]

description_parcours = {
    "DA": "Data Analysis / Excel / bases de l'analyse",
    "BI": "Business Intelligence / Power BI",
    "DS": "Data Science / Python / Machine Learning",
    "IA": "Intelligence Artificielle / LLM / RAG / Agents IA"
}

# Cellule 4 — Fonction de recommandation

In [4]:
# ============================================================
# 4.3. Fonction de recommandation initiale
# ============================================================

def recommander_parcours(row):
    score_da = row["score_initial_DA_pct"]
    score_bi = row["score_initial_BI_pct"]
    score_ds = row["score_initial_DS_pct"]
    score_ia = row["score_initial_IA_pct"]
    score_global = row["score_initial_pourcentage"]
    
    scores = {
        "DA": score_da,
        "BI": score_bi,
        "DS": score_ds,
        "IA": score_ia
    }
    
    domaine_plus_faible = min(scores, key=scores.get)
    
    lacunes = [
        domaine for domaine, score in scores.items()
        if score < SEUIL_DEBUTANT
    ]
    
    # Règle pédagogique progressive
    if score_global < 30:
        parcours = "DA"
        raison = "Niveau global très faible : consolidation des bases recommandée."
    elif score_da < SEUIL_DEBUTANT:
        parcours = "DA"
        raison = "Lacunes en bases d'analyse de données."
    elif score_bi < SEUIL_DEBUTANT:
        parcours = "BI"
        raison = "Bases DA suffisantes, mais lacunes en Business Intelligence."
    elif score_ds < SEUIL_DEBUTANT:
        parcours = "DS"
        raison = "Bases DA/BI suffisantes, mais lacunes en Data Science."
    elif score_ia < SEUIL_DEBUTANT:
        parcours = "IA"
        raison = "Bases suffisantes, mais lacunes en Intelligence Artificielle."
    else:
        parcours = domaine_plus_faible
        raison = "Aucune lacune critique ; recommandation basée sur le domaine le plus faible."
    
    return pd.Series({
        "parcours_recommande": parcours,
        "description_parcours_recommande": description_parcours[parcours],
        "domaine_plus_faible": domaine_plus_faible,
        "lacunes_detectees": ", ".join(lacunes) if lacunes else "Aucune lacune critique",
        "raison_recommandation": raison
    })

# Cellule 5 — Application de la recommandation

In [5]:
# ============================================================
# 4.4. Application de la recommandation
# ============================================================

df_recommandations = df_orientation.copy()

resultats_reco = df_recommandations.apply(
    recommander_parcours,
    axis=1
)

df_recommandations = pd.concat(
    [df_recommandations, resultats_reco],
    axis=1
)

print("Recommandations générées :", df_recommandations.shape)

print("\nRépartition des parcours recommandés :")
display(df_recommandations["parcours_recommande"].value_counts())

print("\nAperçu des recommandations :")
display(df_recommandations[
    [
        "IDENTIFICATION",
        "score_initial_total",
        "score_initial_pourcentage",
        "niveau_initial_global",
        "score_initial_DA_pct",
        "score_initial_BI_pct",
        "score_initial_DS_pct",
        "score_initial_IA_pct",
        "parcours_recommande",
        "domaine_plus_faible",
        "lacunes_detectees",
        "raison_recommandation"
    ]
].head(20))

Recommandations générées : (305, 64)

Répartition des parcours recommandés :


parcours_recommande
DA    265
BI     23
DS     13
IA      4
Name: count, dtype: int64


Aperçu des recommandations :


,IDENTIFICATION,score_initial_total,score_initial_pourcentage,niveau_initial_global,score_initial_DA_pct,score_initial_BI_pct,score_initial_DS_pct,score_initial_IA_pct,parcours_recommande,domaine_plus_faible,lacunes_detectees,raison_recommandation
0,DIEG2026-001,6,30.0,Débutant,0.0,80.0,20.0,20.0,DA,DA,"DA, DS, IA",Lacunes en bases d'analyse de données.
1,DIEG2026-002,5,25.0,Débutant,20.0,40.0,0.0,40.0,DA,DS,"DA, BI, DS, IA",Niveau global très faible : consolidation des ...
2,DIEG2026-003,0,0.0,Débutant,0.0,0.0,0.0,0.0,DA,DA,"DA, BI, DS, IA",Niveau global très faible : consolidation des ...
3,DIEG2026-004,10,50.0,Intermédiaire,80.0,60.0,40.0,20.0,DS,IA,"DS, IA","Bases DA/BI suffisantes, mais lacunes en Data ..."
4,DIEG2026-005,12,60.0,Intermédiaire,60.0,60.0,40.0,80.0,DS,DS,DS,"Bases DA/BI suffisantes, mais lacunes en Data ..."
5,DIEG2026-006,1,5.0,Débutant,0.0,0.0,0.0,20.0,DA,DA,"DA, BI, DS, IA",Niveau global très faible : consolidation des ...
6,DIEG2026-007,8,40.0,Débutant,40.0,20.0,80.0,20.0,DA,BI,"DA, BI, IA",Lacunes en bases d'analyse de données.
7,DIEG2026-008,4,20.0,Débutant,20.0,0.0,20.0,40.0,DA,BI,"DA, BI, DS, IA",Niveau global très faible : consolidation des ...
8,DIEG2026-009,7,35.0,Débutant,60.0,40.0,20.0,20.0,BI,DS,"BI, DS, IA","Bases DA suffisantes, mais lacunes en Business..."
9,DIEG2026-011,4,20.0,Débutant,20.0,0.0,60.0,0.0,DA,BI,"DA, BI, IA",Niveau global très faible : consolidation des ...


Recommandations:
DA : 265
BI : 23
DS : 13
IA : 4

La recommandation est très orientée vers DA, car beaucoup d’apprenants ont un niveau initial faible. C’est logique, mais il faut le documenter.

# Cellule 6 — Analyse et export des recommandations

In [6]:
# ============================================================
# 4.5. Analyse et export des recommandations
# ============================================================

print("===== Répartition des parcours recommandés =====")

repartition_reco = (
    df_recommandations["parcours_recommande"]
    .value_counts()
    .reindex(ordre_parcours, fill_value=0)
)

display(repartition_reco)

repartition_reco_pct = (
    df_recommandations["parcours_recommande"]
    .value_counts(normalize=True)
    .reindex(ordre_parcours, fill_value=0) * 100
).round(2)

display(repartition_reco_pct)

# Tableau résumé
df_repartition_recommandations = pd.DataFrame({
    "parcours_recommande": ordre_parcours,
    "effectif": repartition_reco.values,
    "pourcentage": repartition_reco_pct.values
})

display(df_repartition_recommandations)

# Export
PATH_RECOMMANDATIONS = OUTPUT_DIR / "01_recommandations_initiales.csv"
PATH_REPARTITION_RECOMMANDATIONS = OUTPUT_DIR / "02_repartition_recommandations.csv"

df_recommandations.to_csv(
    PATH_RECOMMANDATIONS,
    index=False,
    encoding="utf-8-sig"
)

df_repartition_recommandations.to_csv(
    PATH_REPARTITION_RECOMMANDATIONS,
    index=False,
    encoding="utf-8-sig"
)

print("Fichier recommandations exporté :", PATH_RECOMMANDATIONS)
print("Fichier répartition exporté :", PATH_REPARTITION_RECOMMANDATIONS)

===== Répartition des parcours recommandés =====


parcours_recommande
DA    265
BI     23
DS     13
IA      4
Name: count, dtype: int64

parcours_recommande
DA    86.89
BI     7.54
DS     4.26
IA     1.31
Name: proportion, dtype: float64

,parcours_recommande,effectif,pourcentage
0,DA,265,86.89
1,BI,23,7.54
2,DS,13,4.26
3,IA,4,1.31


Fichier recommandations exporté : outputs/04_recommandation_parcours/01_recommandations_initiales.csv
Fichier répartition exporté : outputs/04_recommandation_parcours/02_repartition_recommandations.csv


DA : 265 apprenants = 86,89 %
BI : 23 apprenants = 7,54 %
DS : 13 apprenants = 4,26 %
IA : 4 apprenants = 1,31 %

Conclusion courte :

La recommandation oriente fortement vers DA, car la majorité des apprenants ont un niveau initial faible.

# Cellule 7 — Comparaison entre parcours recommandé et parcours réellement suivi

Cette cellule est créer la variable importante :

respect_recommandation

Elle servira plus tard pour comparer :
apprenants ayant appliqué la recommandation IA
vs
apprenants ayant suivi leur propre choix d’apprentissage

In [7]:
# ============================================================
# 4.6. Comparaison entre parcours recommandé et parcours réellement suivi
# ============================================================

# Chargement du dataset prédiction scoré si nécessaire
PATH_PREDICTION_SCORE = Path("outputs/02_scoring_pedagogique/05_dataset_prediction_score.csv")

if "df_prediction_score" not in globals():
    df_prediction_score = pd.read_csv(PATH_PREDICTION_SCORE)

# Colonnes utiles du parcours réellement suivi
colonnes_suivi = [
    "IDENTIFICATION",
    "parcours_test_final",
    "score_final_total",
    "score_final_pourcentage",
    "niveau_final",
    "progression_absolue",
    "reussite_finale"
]

df_suivi = df_prediction_score[colonnes_suivi].copy()

# Fusion recommandation + parcours réellement suivi
df_comparaison_reco_suivi = df_recommandations.merge(
    df_suivi,
    on="IDENTIFICATION",
    how="inner",
    validate="one_to_one"
)

# Respect de la recommandation
df_comparaison_reco_suivi["respect_recommandation"] = (
    df_comparaison_reco_suivi["parcours_recommande"]
    == df_comparaison_reco_suivi["parcours_test_final"]
)

df_comparaison_reco_suivi["groupe_recommandation"] = np.where(
    df_comparaison_reco_suivi["respect_recommandation"],
    "A suivi la recommandation IA",
    "A suivi son propre choix"
)

print("Comparaison recommandation / parcours suivi :", df_comparaison_reco_suivi.shape)

print("\nRespect de la recommandation :")
display(df_comparaison_reco_suivi["respect_recommandation"].value_counts())

print("\nRépartition des groupes :")
display(df_comparaison_reco_suivi["groupe_recommandation"].value_counts())

print("\nMatrice parcours recommandé / parcours suivi :")
matrice_reco_suivi = pd.crosstab(
    df_comparaison_reco_suivi["parcours_recommande"],
    df_comparaison_reco_suivi["parcours_test_final"]
).reindex(index=ordre_parcours, columns=ordre_parcours, fill_value=0)

display(matrice_reco_suivi)

# Export
PATH_COMPARAISON_RECO_SUIVI = OUTPUT_DIR / "03_comparaison_recommandation_parcours_suivi.csv"
PATH_MATRICE_RECO_SUIVI = OUTPUT_DIR / "04_matrice_recommandation_parcours_suivi.csv"

df_comparaison_reco_suivi.to_csv(
    PATH_COMPARAISON_RECO_SUIVI,
    index=False,
    encoding="utf-8-sig"
)

matrice_reco_suivi.to_csv(
    PATH_MATRICE_RECO_SUIVI,
    encoding="utf-8-sig"
)

print("\nFichier comparaison exporté :", PATH_COMPARAISON_RECO_SUIVI)
print("Matrice exportée :", PATH_MATRICE_RECO_SUIVI)

display(df_comparaison_reco_suivi[
    [
        "IDENTIFICATION",
        "score_initial_total",
        "niveau_initial_global",
        "parcours_recommande",
        "parcours_test_final",
        "respect_recommandation",
        "groupe_recommandation",
        "score_final_pourcentage",
        "niveau_final",
        "progression_absolue"
    ]
].head(20))

Comparaison recommandation / parcours suivi : (128, 72)

Respect de la recommandation :


respect_recommandation
False    105
True      23
Name: count, dtype: int64


Répartition des groupes :


groupe_recommandation
A suivi son propre choix        105
A suivi la recommandation IA     23
Name: count, dtype: int64


Matrice parcours recommandé / parcours suivi :


parcours_test_final,DA,BI,DS,IA
parcours_recommande,,,,
DA,18,39,21,36
BI,0,2,3,4
DS,0,0,3,0
IA,0,1,1,0



Fichier comparaison exporté : outputs/04_recommandation_parcours/03_comparaison_recommandation_parcours_suivi.csv
Matrice exportée : outputs/04_recommandation_parcours/04_matrice_recommandation_parcours_suivi.csv


,IDENTIFICATION,score_initial_total,niveau_initial_global,parcours_recommande,parcours_test_final,respect_recommandation,groupe_recommandation,score_final_pourcentage,niveau_final,progression_absolue
0,DIEG2026-003,0,Débutant,DA,IA,False,A suivi son propre choix,93.333333,Avancé,93.333333
1,DIEG2026-007,8,Débutant,DA,IA,False,A suivi son propre choix,80.000000,Avancé,40.000000
2,DIEG2026-008,4,Débutant,DA,BI,False,A suivi son propre choix,100.000000,Avancé,80.000000
3,DIEG2026-009,7,Débutant,BI,IA,False,A suivi son propre choix,93.333333,Avancé,58.333333
4,DIEG2026-012,2,Débutant,DA,IA,False,A suivi son propre choix,53.333333,Intermédiaire,43.333333
5,DIEG2026-025,4,Débutant,DA,DS,False,A suivi son propre choix,86.666667,Avancé,66.666667
6,DIEG2026-026,1,Débutant,DA,BI,False,A suivi son propre choix,60.000000,Intermédiaire,55.000000
7,DIEG2026-027,1,Débutant,DA,BI,False,A suivi son propre choix,60.000000,Intermédiaire,55.000000
8,DIEG2026-028,2,Débutant,DA,BI,False,A suivi son propre choix,80.000000,Avancé,70.000000
9,DIEG2026-029,1,Débutant,DA,IA,False,A suivi son propre choix,73.333333,Avancé,68.333333


Interprétation courte :

128 apprenants ont un test initial + test final.

23 ont suivi la recommandation IA.
105 ont suivi leur propre choix.

Donc :

respect_recommandation = True  → 23 apprenants
respect_recommandation = False → 105 apprenants

La matrice est cohérente : la diagonale donne les apprenants qui ont suivi la recommandation.

==========================================================================================
305 apprenants = ont test initial
→ utilisés pour générer la recommandation initiale
128 apprenants = ont test initial + test final
→ utilisés pour comparer recommandation IA vs parcours réellement suivi

Donc :

Étape recommandation seule : 305 apprenants
Comparaison avec parcours suivi/final : 128 apprenants

# ellule 8 — Comparaison des performances selon respect de la recommandation

In [8]:
# ============================================================
# 4.7. Comparaison des performances selon respect de la recommandation
# ============================================================

print("===== Comparaison des performances =====")

comparaison_performance = (
    df_comparaison_reco_suivi
    .groupby("groupe_recommandation")
    .agg(
        nb_apprenants=("IDENTIFICATION", "count"),
        score_initial_moyen=("score_initial_pourcentage", "mean"),
        score_final_moyen=("score_final_pourcentage", "mean"),
        progression_moyenne=("progression_absolue", "mean"),
        taux_reussite=("reussite_finale", "mean")
    )
    .round(2)
    .reset_index()
)

comparaison_performance["taux_reussite"] = (
    comparaison_performance["taux_reussite"] * 100
).round(2)

display(comparaison_performance)

# Export
PATH_COMPARAISON_PERFORMANCE = OUTPUT_DIR / "05_comparaison_performance_recommandation.csv"

comparaison_performance.to_csv(
    PATH_COMPARAISON_PERFORMANCE,
    index=False,
    encoding="utf-8-sig"
)

print("Fichier exporté :", PATH_COMPARAISON_PERFORMANCE)

===== Comparaison des performances =====


,groupe_recommandation,nb_apprenants,score_initial_moyen,score_final_moyen,progression_moyenne,taux_reussite
0,A suivi la recommandation IA,23,19.13,73.04,53.91,78.0
1,A suivi son propre choix,105,21.81,72.76,50.95,81.0


Fichier exporté : outputs/04_recommandation_parcours/05_comparaison_performance_recommandation.csv


Interprétation courte :

A suivi la recommandation IA :
23 apprenants
Score final moyen : 73,04 %
Progression moyenne : 53,91 points
Taux réussite : 78 %

A suivi son propre choix :
105 apprenants
Score final moyen : 72,76 %
Progression moyenne : 50,95 points
Taux réussite : 81 %

Conclusion provisoire :

Les deux groupes ont des performances finales très proches.
Le groupe recommandation IA a une progression légèrement supérieure.
Mais il faut tester statistiquement avant de conclure.

# Cellule 9 — Test statistique recommandation IA vs choix personnel

In [9]:
# ============================================================
# 4.8. Tests statistiques : recommandation IA vs choix personnel
# ============================================================

from scipy import stats
import numpy as np
import pandas as pd

df_groupe_ia = df_comparaison_reco_suivi[
    df_comparaison_reco_suivi["respect_recommandation"] == True
].copy()

df_groupe_choix = df_comparaison_reco_suivi[
    df_comparaison_reco_suivi["respect_recommandation"] == False
].copy()


def cohen_d(x, y):
    x = pd.Series(x).dropna()
    y = pd.Series(y).dropna()
    
    nx = len(x)
    ny = len(y)
    
    if nx < 2 or ny < 2:
        return np.nan
    
    pooled_std = np.sqrt(
        ((nx - 1) * x.var(ddof=1) + (ny - 1) * y.var(ddof=1))
        / (nx + ny - 2)
    )
    
    if pooled_std == 0:
        return np.nan
    
    return (x.mean() - y.mean()) / pooled_std


variables_a_tester = [
    ("score_final_pourcentage", "Score final (%)"),
    ("progression_absolue", "Progression initiale-finale")
]

resultats_tests = []

for colonne, label in variables_a_tester:
    x = df_groupe_ia[colonne].dropna()
    y = df_groupe_choix[colonne].dropna()
    
    t_stat, p_ttest = stats.ttest_ind(x, y, equal_var=False)
    u_stat, p_mann = stats.mannwhitneyu(x, y, alternative="two-sided")
    
    resultats_tests.append({
        "variable": label,
        "n_recommandation_ia": len(x),
        "n_choix_personnel": len(y),
        "moyenne_recommandation_ia": round(x.mean(), 2),
        "moyenne_choix_personnel": round(y.mean(), 2),
        "difference_moyenne": round(x.mean() - y.mean(), 2),
        "p_value_ttest_welch": round(p_ttest, 4),
        "p_value_mann_whitney": round(p_mann, 4),
        "cohen_d": round(cohen_d(x, y), 4)
    })

df_tests_recommandation = pd.DataFrame(resultats_tests)

print("===== Tests statistiques : recommandation IA vs choix personnel =====")
display(df_tests_recommandation)

# Test du taux de réussite
table_reussite = pd.crosstab(
    df_comparaison_reco_suivi["groupe_recommandation"],
    df_comparaison_reco_suivi["reussite_finale"]
)

print("\n===== Tableau réussite finale =====")
display(table_reussite)

chi2, p_chi2, dof, expected = stats.chi2_contingency(table_reussite)

print("\nTest du Khi-deux sur la réussite finale")
print("chi2 :", round(chi2, 4))
print("p-value :", round(p_chi2, 4))

# Export
PATH_TESTS_RECOMMANDATION = OUTPUT_DIR / "06_tests_statistiques_recommandation.csv"

df_tests_recommandation.to_csv(
    PATH_TESTS_RECOMMANDATION,
    index=False,
    encoding="utf-8-sig"
)

print("\nFichier exporté :", PATH_TESTS_RECOMMANDATION)

===== Tests statistiques : recommandation IA vs choix personnel =====


,variable,n_recommandation_ia,n_choix_personnel,moyenne_recommandation_ia,moyenne_choix_personnel,difference_moyenne,p_value_ttest_welch,p_value_mann_whitney,cohen_d
0,Score final (%),23,105,73.04,72.76,0.28,0.9592,0.9551,0.0119
1,Progression initiale-finale,23,105,53.91,50.95,2.96,0.6986,0.5065,0.1020



===== Tableau réussite finale =====


reussite_finale,0,1
groupe_recommandation,,
A suivi la recommandation IA,5,18
A suivi son propre choix,20,85



Test du Khi-deux sur la réussite finale
chi2 : 0.0
p-value : 0.9964

Fichier exporté : outputs/04_recommandation_parcours/06_tests_statistiques_recommandation.csv


Interprétation courte :

Score final :
73,04 % vs 72,76 %
Différence = +0,28 point
p-value = 0,9592
→ pas significatif

Progression :
53,91 vs 50,95
Différence = +2,96 points
p-value = 0,6986
→ pas significatif

Réussite finale :
78 % vs 81 %
p-value Khi-deux = 0,9964
→ pas significatif

Conclusion :

Les apprenants ayant suivi la recommandation IA ont une progression légèrement supérieure,
mais la différence n’est pas statistiquement significative.

À écrire prudemment dans le mémoire :

Les résultats suggèrent une tendance positive de la recommandation IA sur la progression,
mais cette différence n’est pas statistiquement significative dans l’échantillon étudié.

# Cellule 10 — Rapport automatique de l’Étape 4

In [10]:
# ============================================================
# 4.9. Rapport automatique de l'Étape 4
# ============================================================

import json
from datetime import datetime

def convertir_json(obj):
    if isinstance(obj, dict):
        return {str(k): convertir_json(v) for k, v in obj.items()}
    elif isinstance(obj, list):
        return [convertir_json(v) for v in obj]
    elif isinstance(obj, tuple):
        return list(obj)
    elif hasattr(obj, "item"):
        return obj.item()
    else:
        return obj


rapport_step4 = {
    "etape": "Étape 4 — Recommandation initiale de parcours",
    "date_generation": datetime.now().strftime("%Y-%m-%d %H:%M:%S"),

    "recommandations_sur_305_apprenants": {
        "repartition": df_recommandations["parcours_recommande"].value_counts().to_dict(),
        "repartition_pourcentage": (
            df_recommandations["parcours_recommande"]
            .value_counts(normalize=True)
            .mul(100)
            .round(2)
            .to_dict()
        )
    },

    "comparaison_sur_128_apprenants": {
        "respect_recommandation": df_comparaison_reco_suivi["respect_recommandation"].value_counts().to_dict(),
        "groupes": df_comparaison_reco_suivi["groupe_recommandation"].value_counts().to_dict()
    },

    "performance_par_groupe": comparaison_performance.to_dict(orient="records"),

    "tests_statistiques": df_tests_recommandation.to_dict(orient="records"),

    "test_reussite_finale": {
        "table_reussite": table_reussite.to_dict(),
        "p_value_khi_deux": float(p_chi2)
    },

    "conclusion": (
        "La recommandation initiale oriente majoritairement les apprenants vers le parcours DA, "
        "ce qui est cohérent avec le faible niveau initial global observé. "
        "La comparaison entre les apprenants ayant suivi la recommandation IA et ceux ayant suivi leur propre choix "
        "montre des performances finales proches. La progression moyenne est légèrement supérieure pour le groupe ayant suivi "
        "la recommandation IA, mais les tests statistiques ne montrent pas de différence significative."
    ),

    "prochaine_etape": "Étape 5 — Modélisation prédictive de la performance finale"
}

rapport_step4 = convertir_json(rapport_step4)

# Export JSON
PATH_RAPPORT_STEP4_JSON = OUTPUT_DIR / "rapport_step4_recommandation_parcours.json"

with open(PATH_RAPPORT_STEP4_JSON, "w", encoding="utf-8") as f:
    json.dump(rapport_step4, f, ensure_ascii=False, indent=4)


# Export Markdown
rapport_markdown = f"""# Rapport automatique — Étape 4

## Étape
Étape 4 — Recommandation initiale de parcours

## Date de génération
{rapport_step4["date_generation"]}

## 1. Répartition des parcours recommandés

| Parcours recommandé | Effectif | Pourcentage |
|---|---:|---:|
"""

for _, row in df_repartition_recommandations.iterrows():
    rapport_markdown += f"| {row['parcours_recommande']} | {row['effectif']} | {row['pourcentage']} % |\n"

rapport_markdown += f"""

## 2. Comparaison recommandation IA / choix personnel

| Groupe | Effectif | Score final moyen | Progression moyenne | Taux de réussite |
|---|---:|---:|---:|---:|
"""

for _, row in comparaison_performance.iterrows():
    rapport_markdown += (
        f"| {row['groupe_recommandation']} | {row['nb_apprenants']} | "
        f"{row['score_final_moyen']} % | {row['progression_moyenne']} points | "
        f"{row['taux_reussite']} % |\n"
    )

rapport_markdown += """

## 3. Conclusion

La recommandation initiale oriente majoritairement les apprenants vers le parcours DA, car le niveau initial global est faible.

Les apprenants ayant suivi la recommandation IA présentent une progression moyenne légèrement supérieure, mais la différence n’est pas statistiquement significative.

## 4. Prochaine étape

Étape 5 — Modélisation prédictive de la performance finale.
"""

PATH_RAPPORT_STEP4_MD = OUTPUT_DIR / "rapport_step4_recommandation_parcours.md"

with open(PATH_RAPPORT_STEP4_MD, "w", encoding="utf-8") as f:
    f.write(rapport_markdown)


print("Rapport Étape 4 généré.")
print("JSON :", PATH_RAPPORT_STEP4_JSON)
print("Markdown :", PATH_RAPPORT_STEP4_MD)

print("\nÉtape 4 validée.")
print("Prochaine étape : Étape 5 — Modélisation prédictive de la performance finale")

rapport_step4

Rapport Étape 4 généré.
JSON : outputs/04_recommandation_parcours/rapport_step4_recommandation_parcours.json
Markdown : outputs/04_recommandation_parcours/rapport_step4_recommandation_parcours.md

Étape 4 validée.
Prochaine étape : Étape 5 — Modélisation prédictive de la performance finale


{'etape': 'Étape 4 — Recommandation initiale de parcours',
 'date_generation': '2026-09-20 00:42:01',
 'recommandations_sur_305_apprenants': {'repartition': {'DA': 265,
   'BI': 23,
   'DS': 13,
   'IA': 4},
  'repartition_pourcentage': {'DA': 86.89,
   'BI': 7.54,
   'DS': 4.26,
   'IA': 1.31}},
 'comparaison_sur_128_apprenants': {'respect_recommandation': {'False': 105,
   'True': 23},
  'groupes': {'A suivi son propre choix': 105,
   'A suivi la recommandation IA': 23}},
 'performance_par_groupe': [{'groupe_recommandation': 'A suivi la recommandation IA',
   'nb_apprenants': 23,
   'score_initial_moyen': 19.13,
   'score_final_moyen': 73.04,
   'progression_moyenne': 53.91,
   'taux_reussite': 78.0},
  {'groupe_recommandation': 'A suivi son propre choix',
   'nb_apprenants': 105,
   'score_initial_moyen': 21.81,
   'score_final_moyen': 72.76,
   'progression_moyenne': 50.95,
   'taux_reussite': 81.0}],
 'tests_statistiques': [{'variable': 'Score final (%)',
   'n_recommandation_ia':

Résumé :

Recommandations sur 305 apprenants :
DA : 265 soit 86,89 %
BI : 23 soit 7,54 %
DS : 13 soit 4,26 %
IA : 4 soit 1,31 %

Comparaison sur les 128 apprenants ayant test initial + test final :

A suivi la recommandation IA : 23
A suivi son propre choix : 105

Performance :

Recommandation IA :
score final moyen = 73,04 %
progression moyenne = 53,91 points
taux réussite = 78 %

Choix personnel :
score final moyen = 72,76 %
progression moyenne = 50,95 points
taux réussite = 81 %

Conclusion :

La recommandation IA donne une progression légèrement meilleure,
mais la différence n’est pas statistiquement significative.

#  Cellule 11 — Recommandation de contenus par parcours et niveau

La recommandation de contenu vient juste après la recommandation de parcours.

Étape 4 — Recommandation initiale de parcours
Étape 4 bis — Recommandation de contenus par parcours et par niveau
Étape 4 ter — Ajustement des contenus après le test intermédiaire

Pourquoi ?

Parce que pour recommander un contenu, il faut connaître :

1. le parcours recommandé : DA / BI / DS / IA ;
2. le niveau ou cluster : Débutant / Intermédiaire / Avancé ;
3. les lacunes détectées dans le test initial ;
4. les lacunes détectées dans le test intermédiaire.

La logique devient donc :

Test initial
→ scores par domaine
→ niveau / cluster
→ parcours recommandé
→ contenus recommandés au début du parcours

Puis :

Test intermédiaire
→ détection des lacunes restantes
→ ajustement ou remédiation des contenus
→ exercices ciblés avant le test final

Exemple :

Parcours recommandé : DA
Niveau : Débutant
Lacunes initiales : Excel, nettoyage des données, TCD
Contenus recommandés :
- bases Excel ;
- nettoyage de données ;
- tableaux croisés dynamiques ;
- exercices simples.

Après test intermédiaire :
Lacune détectée : jointure ou modèle de données
Contenus de remédiation :
- rappel sur les clés de jointure ;
- exercices sur Power Query ;
- cas pratique guidé ;
- mini-évaluation avant le test final.

Pour le mémoire, il faut donc ajouter :

Étape 4 bis — Recommandation de contenus pédagogiques personnalisés
Étape 4 ter — Remédiation à partir des lacunes du test intermédiaire

Sortie attendue :

recommandations_contenus.csv
recommandations_remediation_intermediaire.csv

Colonnes possibles :

IDENTIFICATION
parcours_recommande
niveau_initial_global
cluster_initial
lacunes_initiales_detectees
contenu_recommande
type_contenu
priorite
raison_recommandation

Et pour la remédiation :

IDENTIFICATION
parcours_suivi
niveau_intermediaire
lacunes_intermediaires_detectees
contenu_remediation
type_remediation
priorite
raison_remediation

Conclusion courte :

La recommandation de parcours dit : quel chemin suivre.
La recommandation de contenu dit : quoi apprendre dans ce chemin selon le niveau.
La remédiation intermédiaire dit : quoi corriger pendant la formation avant le test final.

In [11]:
# ============================================================
# 4.10. Étape 4 bis — Recommandation de contenus par parcours et par niveau
# ============================================================

catalogue_contenus = {
    "DA": {
        "Débutant": [
            "Bases Excel",
            "Nettoyage simple des données",
            "Tableaux croisés dynamiques",
            "Graphiques simples"
        ],
        "Intermédiaire": [
            "Power Query",
            "Modèle de données Excel",
            "Jointures et relations",
            "Tableaux de bord simples"
        ],
        "Avancé": [
            "Power Pivot",
            "Mesures DAX de base",
            "Dashboard analytique",
            "Cas pratique complet DA"
        ]
    },
    "BI": {
        "Débutant": [
            "Introduction Power BI",
            "Importation des données",
            "Power Query débutant",
            "Visualisations simples"
        ],
        "Intermédiaire": [
            "Modèle en étoile",
            "Relations entre tables",
            "Mesures DAX",
            "Filtres et segments"
        ],
        "Avancé": [
            "DAX avancé",
            "Row-Level Security",
            "Optimisation du modèle",
            "Dashboard décisionnel"
        ]
    },
    "DS": {
        "Débutant": [
            "Bases Python",
            "Pandas pour CSV",
            "Nettoyage des données",
            "Statistiques descriptives"
        ],
        "Intermédiaire": [
            "Feature engineering",
            "Train/test split",
            "Classification et régression",
            "Évaluation des modèles"
        ],
        "Avancé": [
            "Random Forest",
            "XGBoost",
            "Optimisation des hyperparamètres",
            "Déploiement modèle avec FastAPI"
        ]
    },
    "IA": {
        "Débutant": [
            "Bases des LLM",
            "Prompt engineering",
            "System prompt",
            "Sorties structurées JSON"
        ],
        "Intermédiaire": [
            "RAG",
            "Embeddings",
            "Vector database",
            "Function calling"
        ],
        "Avancé": [
            "Agents IA",
            "MCP",
            "Workflow n8n",
            "Architecture IA complète"
        ]
    }
}

def recommander_contenus(row):
    parcours = row["parcours_recommande"]
    niveau = row["niveau_initial_global"]
    
    contenus = catalogue_contenus.get(parcours, {}).get(niveau, [])
    
    if niveau == "Débutant":
        priorite = "Haute"
    elif niveau == "Intermédiaire":
        priorite = "Moyenne"
    else:
        priorite = "Renforcement"
    
    return pd.Series({
        "contenus_recommandes": " ; ".join(contenus),
        "type_contenu": "Cours + exercices + mini-évaluation",
        "priorite_contenu": priorite,
        "raison_contenu": f"Contenus proposés selon le parcours {parcours} et le niveau {niveau}."
    })

df_recommandations_contenus = df_recommandations.copy()

resultats_contenus = df_recommandations_contenus.apply(
    recommander_contenus,
    axis=1
)

df_recommandations_contenus = pd.concat(
    [df_recommandations_contenus, resultats_contenus],
    axis=1
)

PATH_RECOMMANDATIONS_CONTENUS = OUTPUT_DIR / "07_recommandations_contenus_parcours_niveau.csv"

df_recommandations_contenus.to_csv(
    PATH_RECOMMANDATIONS_CONTENUS,
    index=False,
    encoding="utf-8-sig"
)

print("Recommandations de contenus générées :", df_recommandations_contenus.shape)
print("Fichier exporté :", PATH_RECOMMANDATIONS_CONTENUS)

display(df_recommandations_contenus[
    [
        "IDENTIFICATION",
        "niveau_initial_global",
        "parcours_recommande",
        "lacunes_detectees",
        "contenus_recommandes",
        "priorite_contenu",
        "raison_contenu"
    ]
].head(20))

Recommandations de contenus générées : (305, 68)
Fichier exporté : outputs/04_recommandation_parcours/07_recommandations_contenus_parcours_niveau.csv


,IDENTIFICATION,niveau_initial_global,parcours_recommande,lacunes_detectees,contenus_recommandes,priorite_contenu,raison_contenu
0,DIEG2026-001,Débutant,DA,"DA, DS, IA",Bases Excel ; Nettoyage simple des données ; T...,Haute,Contenus proposés selon le parcours DA et le n...
1,DIEG2026-002,Débutant,DA,"DA, BI, DS, IA",Bases Excel ; Nettoyage simple des données ; T...,Haute,Contenus proposés selon le parcours DA et le n...
2,DIEG2026-003,Débutant,DA,"DA, BI, DS, IA",Bases Excel ; Nettoyage simple des données ; T...,Haute,Contenus proposés selon le parcours DA et le n...
3,DIEG2026-004,Intermédiaire,DS,"DS, IA",Feature engineering ; Train/test split ; Class...,Moyenne,Contenus proposés selon le parcours DS et le n...
4,DIEG2026-005,Intermédiaire,DS,DS,Feature engineering ; Train/test split ; Class...,Moyenne,Contenus proposés selon le parcours DS et le n...
5,DIEG2026-006,Débutant,DA,"DA, BI, DS, IA",Bases Excel ; Nettoyage simple des données ; T...,Haute,Contenus proposés selon le parcours DA et le n...
6,DIEG2026-007,Débutant,DA,"DA, BI, IA",Bases Excel ; Nettoyage simple des données ; T...,Haute,Contenus proposés selon le parcours DA et le n...
7,DIEG2026-008,Débutant,DA,"DA, BI, DS, IA",Bases Excel ; Nettoyage simple des données ; T...,Haute,Contenus proposés selon le parcours DA et le n...
8,DIEG2026-009,Débutant,BI,"BI, DS, IA",Introduction Power BI ; Importation des donnée...,Haute,Contenus proposés selon le parcours BI et le n...
9,DIEG2026-011,Débutant,DA,"DA, BI, IA",Bases Excel ; Nettoyage simple des données ; T...,Haute,Contenus proposés selon le parcours DA et le n...


# Cellule 12 — Ajustement après test intermédiaire

In [12]:
# ============================================================
# 4.11. Étape 4 ter — Ajustement des contenus après test intermédiaire
# VERSION CORRIGÉE
# ============================================================

# Charger les fichiers nécessaires si les variables n'existent pas
INPUT_SCORE_DIR = Path("outputs/02_scoring_pedagogique")

PATH_SCORES_INTERMEDIAIRES = INPUT_SCORE_DIR / "02_scores_tests_intermediaires.csv"
PATH_CORRECTION_MI = INPUT_SCORE_DIR / "template_correction_tests_intermediaires.csv"

if "df_scores_intermediaires" not in globals():
    df_scores_intermediaires = pd.read_csv(PATH_SCORES_INTERMEDIAIRES)

if "df_correction_mi" not in globals():
    df_correction_mi = pd.read_csv(PATH_CORRECTION_MI)

print("Scores intermédiaires :", df_scores_intermediaires.shape)
print("Barème intermédiaire :", df_correction_mi.shape)


# Colonnes de score des questions intermédiaires
colonnes_score_mi = [
    col for col in df_scores_intermediaires.columns
    if col.startswith("score_mi_Q")
]

print("Colonnes scores MI détectées :", len(colonnes_score_mi))


# Mapping question standardisée -> question originale
mapping_questions_mi = (
    df_correction_mi
    .set_index(["parcours", "question_standardisee"])["question_originale"]
    .to_dict()
)


def detecter_lacunes_intermediaires(row):
    parcours = row["parcours_test_intermediaire"]
    lacunes = []
    
    for col_score in colonnes_score_mi:
        question = col_score.replace("score_", "")
        score = row[col_score]
        
        if score == 0:
            question_originale = mapping_questions_mi.get(
                (parcours, question),
                question
            )
            lacunes.append(question_originale)
    
    return lacunes


def recommander_remediation(row):
    lacunes = detecter_lacunes_intermediaires(row)
    niveau = row["niveau_intermediaire"]
    parcours = row["parcours_test_intermediaire"]
    
    if len(lacunes) == 0:
        return pd.Series({
            "lacunes_intermediaires_detectees": "Aucune lacune majeure",
            "contenu_remediation": "Renforcement avancé et cas pratique complet",
            "type_remediation": "Approfondissement",
            "priorite_remediation": "Faible",
            "raison_remediation": "Aucune erreur détectée au test intermédiaire."
        })
    
    if niveau == "Débutant":
        priorite = "Haute"
        type_remediation = "Remédiation obligatoire"
    elif niveau == "Intermédiaire":
        priorite = "Moyenne"
        type_remediation = "Remédiation ciblée"
    else:
        priorite = "Faible"
        type_remediation = "Renforcement ciblé"
    
    contenu = (
        f"Revoir les notions non maîtrisées du parcours {parcours}, "
        "faire des exercices ciblés, puis une mini-évaluation avant le test final."
    )
    
    return pd.Series({
        "lacunes_intermediaires_detectees": " | ".join(lacunes),
        "contenu_remediation": contenu,
        "type_remediation": type_remediation,
        "priorite_remediation": priorite,
        "raison_remediation": f"{len(lacunes)} lacune(s) détectée(s) au test intermédiaire."
    })


df_remediation_intermediaire = df_scores_intermediaires.copy()

resultats_remediation = df_remediation_intermediaire.apply(
    recommander_remediation,
    axis=1
)

df_remediation_intermediaire = pd.concat(
    [df_remediation_intermediaire, resultats_remediation],
    axis=1
)


PATH_REMEDIATION = OUTPUT_DIR / "08_remediation_apres_test_intermediaire.csv"

df_remediation_intermediaire.to_csv(
    PATH_REMEDIATION,
    index=False,
    encoding="utf-8-sig"
)

print("Remédiation intermédiaire générée :", df_remediation_intermediaire.shape)
print("Fichier exporté :", PATH_REMEDIATION)

print("\nRépartition des priorités de remédiation :")
display(df_remediation_intermediaire["priorite_remediation"].value_counts())

display(df_remediation_intermediaire[
    [
        "IDENTIFICATION",
        "parcours_test_intermediaire",
        "score_intermediaire_total",
        "niveau_intermediaire",
        "lacunes_intermediaires_detectees",
        "contenu_remediation",
        "priorite_remediation"
    ]
].head(20))

Scores intermédiaires : (196, 37)
Barème intermédiaire : (60, 7)
Colonnes scores MI détectées : 15
Remédiation intermédiaire générée : (196, 42)
Fichier exporté : outputs/04_recommandation_parcours/08_remediation_apres_test_intermediaire.csv

Répartition des priorités de remédiation :


priorite_remediation
Faible     108
Moyenne     54
Haute       34
Name: count, dtype: int64

,IDENTIFICATION,parcours_test_intermediaire,score_intermediaire_total,niveau_intermediaire,lacunes_intermediaires_detectees,contenu_remediation,priorite_remediation
0,DIEG2026-463,DA,8,Intermédiaire,"Parmi les 4 piliers de la qualite des donnees,...",Revoir les notions non maîtrisées du parcours ...,Moyenne
1,DIEG2026-636,DA,13,Avancé,"Dans le processus ETL, l'etape Transform corre...",Revoir les notions non maîtrisées du parcours ...,Faible
2,DIEG2026-633,DA,10,Intermédiaire,Pourquoi Power Query affiche-t-il uniquement ~...,Revoir les notions non maîtrisées du parcours ...,Moyenne
3,DIEG2026-442,DA,8,Intermédiaire,"Parmi les 4 piliers de la qualite des donnees,...",Revoir les notions non maîtrisées du parcours ...,Moyenne
4,DIEG2026-055,DA,8,Intermédiaire,"Parmi les 4 piliers de la qualite des donnees,...",Revoir les notions non maîtrisées du parcours ...,Moyenne
5,DIEG2026-601,DA,11,Avancé,Pourquoi Power Query affiche-t-il uniquement ~...,Revoir les notions non maîtrisées du parcours ...,Faible
6,DIEG2026-467,DA,15,Avancé,Aucune lacune majeure,Renforcement avancé et cas pratique complet,Faible
7,DIEG2026-417,DA,13,Avancé,Une direction DIR (avec un espace final) dans ...,Revoir les notions non maîtrisées du parcours ...,Faible
8,DIEG2026-117,DA,9,Intermédiaire,Qu'exprime le principe GIGO en Data Analysis ?...,Revoir les notions non maîtrisées du parcours ...,Moyenne
9,DIEG2026-327,DA,13,Avancé,Pourquoi Power Query affiche-t-il uniquement ~...,Revoir les notions non maîtrisées du parcours ...,Faible


Résumé :

Étape 4 bis :
Recommandations de contenus générées pour 305 apprenants.
Fichier : 07_recommandations_contenus_parcours_niveau.csv
Étape 4 ter :
Remédiation intermédiaire générée pour 196 apprenants.
Fichier : 08_remediation_apres_test_intermediaire.csv

Priorités de remédiation :

Faible : 108
Moyenne : 54
Haute : 34

Interprétation courte :

Étape 4 bis = quoi apprendre au début selon parcours + niveau.
Étape 4 ter = quoi corriger après le test intermédiaire selon les erreurs.

# Cellule 13 — Rapport complémentaire Étape 4 bis / 4 ter

In [13]:
# ============================================================
# 4.12. Rapport complémentaire — contenus et remédiation
# ============================================================

import json
from datetime import datetime

rapport_step4_complement = {
    "etape": "Étape 4 bis et 4 ter — Recommandation de contenus et remédiation",
    "date_generation": datetime.now().strftime("%Y-%m-%d %H:%M:%S"),

    "recommandation_contenus": {
        "nb_apprenants": int(len(df_recommandations_contenus)),
        "fichier": str(PATH_RECOMMANDATIONS_CONTENUS),
        "objectif": "Proposer des contenus pédagogiques selon le parcours recommandé et le niveau initial."
    },

    "remediation_intermediaire": {
        "nb_apprenants": int(len(df_remediation_intermediaire)),
        "fichier": str(PATH_REMEDIATION),
        "repartition_priorites": df_remediation_intermediaire["priorite_remediation"].value_counts().to_dict(),
        "objectif": "Adapter les contenus après le test intermédiaire selon les lacunes restantes."
    },

    "conclusion": (
        "La recommandation de contenus complète la recommandation de parcours. "
        "Elle permet de proposer des contenus adaptés au niveau initial de l'apprenant. "
        "La remédiation intermédiaire permet ensuite de corriger les lacunes détectées pendant la formation."
    ),

    "prochaine_etape": "Étape 5 — Modélisation prédictive de la performance finale"
}

PATH_RAPPORT_COMPLEMENT_JSON = OUTPUT_DIR / "rapport_step4bis_4ter_contenus_remediation.json"

with open(PATH_RAPPORT_COMPLEMENT_JSON, "w", encoding="utf-8") as f:
    json.dump(rapport_step4_complement, f, ensure_ascii=False, indent=4)


rapport_markdown = f"""# Rapport complémentaire — Étape 4 bis et 4 ter

## 1. Recommandation de contenus

Nombre d'apprenants concernés : {len(df_recommandations_contenus)}

Objectif : proposer des contenus pédagogiques selon le parcours recommandé et le niveau initial.

Fichier généré :

{PATH_RECOMMANDATIONS_CONTENUS}

## 2. Remédiation après test intermédiaire

Nombre d'apprenants concernés : {len(df_remediation_intermediaire)}

Objectif : ajuster les contenus après le test intermédiaire selon les lacunes restantes.

Fichier généré :

{PATH_REMEDIATION}

## 3. Répartition des priorités de remédiation

"""

for priorite, effectif in df_remediation_intermediaire["priorite_remediation"].value_counts().items():
    rapport_markdown += f"- {priorite} : {effectif}\n"

rapport_markdown += """

## 4. Conclusion

La recommandation de parcours indique quel chemin suivre.

La recommandation de contenus indique quoi apprendre dans ce parcours selon le niveau.

La remédiation intermédiaire indique quoi corriger pendant la formation avant le test final.

## 5. Prochaine étape

Étape 5 — Modélisation prédictive de la performance finale.
"""

PATH_RAPPORT_COMPLEMENT_MD = OUTPUT_DIR / "rapport_step4bis_4ter_contenus_remediation.md"

with open(PATH_RAPPORT_COMPLEMENT_MD, "w", encoding="utf-8") as f:
    f.write(rapport_markdown)

print("Rapport complémentaire généré.")
print("JSON :", PATH_RAPPORT_COMPLEMENT_JSON)
print("Markdown :", PATH_RAPPORT_COMPLEMENT_MD)

print("\nÉtape 4 bis et 4 ter validées.")
print("Prochaine étape : Étape 5 — Modélisation prédictive de la performance finale")

rapport_step4_complement

Rapport complémentaire généré.
JSON : outputs/04_recommandation_parcours/rapport_step4bis_4ter_contenus_remediation.json
Markdown : outputs/04_recommandation_parcours/rapport_step4bis_4ter_contenus_remediation.md

Étape 4 bis et 4 ter validées.
Prochaine étape : Étape 5 — Modélisation prédictive de la performance finale


{'etape': 'Étape 4 bis et 4 ter — Recommandation de contenus et remédiation',
 'date_generation': '2026-09-20 00:42:02',
 'recommandation_contenus': {'nb_apprenants': 305,
  'fichier': 'outputs/04_recommandation_parcours/07_recommandations_contenus_parcours_niveau.csv',
  'objectif': 'Proposer des contenus pédagogiques selon le parcours recommandé et le niveau initial.'},
 'remediation_intermediaire': {'nb_apprenants': 196,
  'fichier': 'outputs/04_recommandation_parcours/08_remediation_apres_test_intermediaire.csv',
  'repartition_priorites': {'Faible': 108, 'Moyenne': 54, 'Haute': 34},
  'objectif': 'Adapter les contenus après le test intermédiaire selon les lacunes restantes.'},
 'conclusion': "La recommandation de contenus complète la recommandation de parcours. Elle permet de proposer des contenus adaptés au niveau initial de l'apprenant. La remédiation intermédiaire permet ensuite de corriger les lacunes détectées pendant la formation.",
 'prochaine_etape': 'Étape 5 — Modélisatio

Résumé :

Étape 4 bis :
305 apprenants ont reçu une recommandation de contenus.
Fichier généré :
07_recommandations_contenus_parcours_niveau.csv
Étape 4 ter :
196 apprenants ont reçu une remédiation après test intermédiaire.
Fichier généré :
08_remediation_apres_test_intermediaire.csv

Priorités de remédiation :

Faible : 108
Moyenne : 54
Haute : 34

Conclusion :

Parcours recommandé = quel chemin suivre.
Contenus recommandés = quoi apprendre au début.
Remédiation intermédiaire = quoi corriger pendant la formation.